### DISCUSSION TITLE
Insiders: Getting material-mkdocs-insiders into requirements.txt (CI, readthedocs)

### DISCUSSION THREAD

**Discussion by rfay** [squidfunk/mkdocs-material#4074] [General]
I'm able to access material-mkdocs-insiders just fine using pip using the [instructions](https://squidfunk.github.io/mkdocs-material/insiders/getting-started/?h=insiders#with-pip), but was thinking it would be best practice to incorporate it into requirements.txt. Is there a way to do that with an exotic URL like `git+https://${GH_TOKEN}@github.com/squidfunk/mkdocs-material-insiders.git` ? Or should it just be installed separately with the pip command? And how is readthedocs going to be able to access it?

**Comment by squidfunk** [comment]
Jup, it's definitely best practice, but sadly, `pip`'s `requirements.txt` doesn't allow for environment variables. @StarfallProjects did a [write-up](https://www.starfallprojects.co.uk/projects/deploy-host-docs/deploy-mkdocs-material-digitalocean-app-platform/#deployment) on how to deploy Insiders to different providers, but I'm not sure whether readthedocs is included. There are probably other cases where readthedocs needs to build from private repositories with secrets, so maybe ask the maintainers of readthedocs?

**Reply by rfay** [reply]
Thanks. I asked in https://github.com/readthedocs/readthedocs.org/issues/9388

**Reply by rfay** [reply]
I can't imagine you'd approve, but the mkdocs-material-insiders could be incorporated into my project and then installed by pip from local, true. But that's technically feasible?

**Reply by rfay** [reply]
The answer is https://docs.readthedocs.io/en/stable/guides/private-python-packages.html - at least in the context of readthedocs, you can include environment variables in the requirements.txt, and they give the example how to do it. I used `git+https://${MKDOCS_GITHUB_USER}:${MKDOCS_GITHUB_TOKEN}@github.com/squidfunk/mkdocs-material-insiders.git@8.3.8-insiders-4.19.1`

**Reply by squidfunk** [reply]
Perfect, sounds like a good solution! Note that I'd kindly ask you not to share the Insiders code, as it would counteract the whole release strategy, making my work on this project not possible anymore.

**Reply by rfay** [reply]
Yes, totally understood :) Doesn't make any sense to share it. And of course, with the solution here it's not even a reasonable thing.

**Reply by NicolasMassart** [reply]
@rfay I have the same need. Have you had success already with this way to use insider? Just to confirm that it is the right way...
Thanks!

**Reply by NicolasMassart** [reply]
It says only for RTD for business. Have you tried with the RTD.org?

**Reply by rfay** [reply]
Yes, it worked fine with readthedocs.org - they process it. As a matter of fact, pip3 does in fact process environment variables, and this all works out fine.

**Reply by NicolasMassart** [reply]
Thanks a lot!

**Reply by NicolasMassart** [reply]
I have to check if we have a Github machine user to generate a token now...

**Reply by NicolasMassart** [reply]
I confirm this works like a charm with user token and RTD requirements.txt
